`client.memory_dump_test` failing on recent Linux kernels

<details>
<summary>Full log</summary>

```
ctest -V -R client.memory_dump_test
UpdateCTestConfiguration  from :/home/binwang/projects/dynamorio/build/DartConfiguration.tcl
Parse Config file:/home/binwang/projects/dynamorio/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/binwang/projects/dynamorio/build/DartConfiguration.tcl
Parse Config file:/home/binwang/projects/dynamorio/build/DartConfiguration.tcl
Test project /home/binwang/projects/dynamorio/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 199
    Start 199: code_api|client.memory_dump_test

199: Test command: /usr/bin/cmake "-D" "precmd=rm@/home/binwang/projects/dynamorio/build/logs/simple_app.*.elf" "-D" "cmd=/home/binwang/projects/dynamorio/build/bin64/drrun@-s@90@-quiet@-debug@-killpg@-stderr_mask@0xC@-dumpcore_mask@0@-code_api@-c@/home/binwang/projects/dynamorio/build/suite/tests/bin/libclient.memory_dump_test.dll.so@--@/home/binwang/projects/dynamorio/build/suite/tests/bin/simple_app" "-D" "postcmd=/usr/bin/readelf@-a@/home/binwang/projects/dynamorio/build/logs/simple_app.*.elf" "-D" "postcmd2=" "-D" "postcmd3=" "-D" "postcmd4=" "-D" "failok=" "-D" "cmp=/home/binwang/projects/dynamorio/build/suite/tests/memory_dump_test.expect" "-D" "code=" "-D" "capture=" "-D" "ignore_matching_lines=" "-P" "/home/binwang/projects/dynamorio/suite/tests/runmulti.cmake"
199: Working Directory: /home/binwang/projects/dynamorio/build/suite/tests
199: Test timeout computed to be: 90
199: Running precmd |rm;/home/binwang/projects/dynamorio/build/logs/simple_app.197233.0c35df0a6084.00000000.elf|
199: Running cmd |/home/binwang/projects/dynamorio/build/bin64/drrun;-s;90;-quiet;-debug;-killpg;-stderr_mask;0xC;-dumpcore_mask;0;-code_api;-c;/home/binwang/projects/dynamorio/build/suite/tests/bin/libclient.memory_dump_test.dll.so;--;/home/binwang/projects/dynamorio/build/suite/tests/bin/simple_app|
199: Running postcmd |/usr/bin/readelf;-a;/home/binwang/projects/dynamorio/build/logs/simple_app.204832.0c35df0c4717.00000000.elf|
199: CMake Error at /home/binwang/projects/dynamorio/suite/tests/runmulti.cmake:72 (message):
199:   output |thank you for testing memory dump
199: 
199:   thread init
199: 
199:   nudge delivered 1
199: 
199:   <Failed to write the requested memory content into the core dump file.>
199: 
199:   <section: [vvar_vclock], prot: 1, length: 8192, written: -14
199: 
199:   >
199: 
199:   Error: failed to create memory dump.
199: 
199:   Hello, world!
199: 
199:   readelf: Error: Reading 3136 bytes extends past end of file for section
199:   headers
199: 
199:   readelf: Error: Section headers are not available!
199: 
199:   ELF Header:
199: 
199:     Magic:   7f 45 4c 46 02 01 01 03 00 00 00 00 00 00 00 00 
199:     Class:                             ELF64
199:     Data:                              2's complement, little endian
199:     Version:                           1 (current)
199:     OS/ABI:                            UNIX - GNU
199:     ABI Version:                       0
199:     Type:                              CORE (Core file)
199:     Machine:                           Advanced Micro Devices X86-64
199:     Version:                           0x1
199:     Entry point address:               0x7fa9d251c540
199:     Start of program headers:          64 (bytes into file)
199:     Start of section headers:          10502452 (bytes into file)
199:     Flags:                             0x0
199:     Size of this header:               64 (bytes)
199:     Size of program headers:           56 (bytes)
199:     Number of program headers:         49
199:     Size of section headers:           64 (bytes)
199:     Number of section headers:         49
199:     Section header string table index: 48
199: 
199:   
199: 
199:   Program Headers:
199: 
199:     Type           Offset             VirtAddr           PhysAddr
199:                    FileSiz            MemSiz              Flags  Align
199:     NOTE           0x0000000000000af8 0x0000000000000000 0x0000000000000000
199:                    0x0000000000000378 0x0000000000000378         0x2
199:     LOAD           0x0000000000001000 0x00007fa78dff8000 0x00007fa78dff8000
199:                    0x0000000000040000 0x0000000000040000  RW     0x1000
199:     LOAD           0x0000000000041000 0x00007fa78e039000 0x00007fa78e039000
199:                    0x0000000000006000 0x0000000000006000  RW     0x1000
199:     LOAD           0x0000000000047000 0x00007fa78e041000 0x00007fa78e041000
199:                    0x0000000000004000 0x0000000000004000  RW     0x1000
199:     LOAD           0x000000000004b000 0x00007fa78e049000 0x00007fa78e049000
199:                    0x0000000000006000 0x0000000000006000  RW     0x1000
199:     LOAD           0x0000000000051000 0x00007fa78e051000 0x00007fa78e051000
199:                    0x0000000000018000 0x0000000000018000  RW     0x1000
199:     LOAD           0x0000000000069000 0x00007fa78e06b000 0x00007fa78e06b000
199:                    0x000000000000e000 0x000000000000e000  RW     0x1000
199:     LOAD           0x0000000000077000 0x00007fa78e07b000 0x00007fa78e07b000
199:                    0x000000000000c000 0x000000000000c000  RW     0x1000
199:     LOAD           0x0000000000083000 0x00007fa78e089000 0x00007fa78e089000
199:                    0x0000000000003000 0x0000000000003000  RW     0x1000
199:     LOAD           0x0000000000086000 0x00007fa78e097000 0x00007fa78e097000
199:                    0x0000000000001000 0x0000000000001000  RW     0x1000
199:     LOAD           0x0000000000087000 0x00007fa78e09a000 0x00007fa78e09a000
199:                    0x0000000000002000 0x0000000000002000  RW     0x1000
199:     LOAD           0x0000000000089000 0x00007fa78e09e000 0x00007fa78e09e000
199:                    0x000000000000e000 0x000000000000e000  RW     0x1000
199:     LOAD           0x0000000000097000 0x00007fa78e0ae000 0x00007fa78e0ae000
199:                    0x0000000000004000 0x0000000000004000  RW     0x1000
199:     LOAD           0x000000000009b000 0x00007fa78e0b6000 0x00007fa78e0b6000
199:                    0x0000000000002000 0x0000000000002000  RW     0x1000
199:     LOAD           0x000000000009d000 0x00007fa78e0bb000 0x00007fa78e0bb000
199:                    0x0000000000001000 0x0000000000001000  RW     0x1000
199:     LOAD           0x000000000009e000 0x00007fa78e0c3000 0x00007fa78e0c3000
199:                    0x0000000000008000 0x0000000000008000  RW     0x1000
199:     LOAD           0x00000000000a6000 0x00007fa78e0cd000 0x00007fa78e0cd000
199:                    0x0000000000018000 0x0000000000018000  RW     0x1000
199:     LOAD           0x00000000000be000 0x00007fa98dff9000 0x00007fa98dff9000
199:                    0x0000000000008000 0x0000000000008000  RW     0x1000
199:     LOAD           0x00000000000c6000 0x00007fa98e001000 0x00007fa98e001000
199:                    0x0000000000001000 0x0000000000001000  R      0x1000
199:     LOAD           0x00000000000c7000 0x00007fa98e002000 0x00007fa98e002000
199:                    0x0000000000001000 0x0000000000001000  R E    0x1000
199:     LOAD           0x00000000000c8000 0x00007fa98e003000 0x00007fa98e003000
199:                    0x0000000000001000 0x0000000000001000  R      0x1000
199:     LOAD           0x00000000000c9000 0x00007fa98e004000 0x00007fa98e004000
199:                    0x0000000000002000 0x0000000000002000  RW     0x1000
199:     LOAD           0x00000000000cb000 0x00007fa98e008000 0x00007fa98e008000
199:                    0x0000000000001000 0x0000000000001000  RW     0x1000
199:     LOAD           0x00000000000cc000 0x00007fa98e010000 0x00007fa98e010000
199:                    0x0000000000001000 0x0000000000001000  RWE    0x1000
199:     LOAD           0x00000000000cd000 0x00007fa98e011000 0x00007fa98e011000
199:                    0x0000000000003000 0x0000000000003000  R E    0x1000
199:     LOAD           0x00000000000d0000 0x00007fa98e014000 0x00007fa98e014000
199:                    0x0000000000001000 0x0000000000001000  RWE    0x1000
199:     LOAD           0x00000000000d1000 0x00007fa98e017000 0x00007fa98e017000
199:                    0x0000000000001000 0x0000000000001000  RWE    0x1000
199:     LOAD           0x00000000000d2000 0x00007fa98e027000 0x00007fa98e027000
199:                    0x0000000000001000 0x0000000000001000  RWE    0x1000
199:     LOAD           0x00000000000d3000 0x00007fa98e037000 0x00007fa98e037000
199:                    0x0000000000001000 0x0000000000001000  RWE    0x1000
199:     LOAD           0x00000000000d4000 0x00007fa98e03a000 0x00007fa98e03a000
199:                    0x0000000000001000 0x0000000000001000  RWE    0x1000
199:     LOAD           0x00000000000d5000 0x00007fa98e03d000 0x00007fa98e03d000
199:                    0x0000000000001000 0x0000000000001000  RW     0x1000
199:     LOAD           0x00000000000d6000 0x00007fa9cdff9000 0x00007fa9cdff9000
199:                    0x0000000000001000 0x0000000000001000  R      0x1000
199:     LOAD           0x00000000000d7000 0x00007fa9cdffa000 0x00007fa9cdffa000
199:                    0x0000000000002000 0x0000000000002000  R E    0x1000
199:     LOAD           0x00000000000d9000 0x00007fa9cdffc000 0x00007fa9cdffc000
199:                    0x0000000000001000 0x0000000000001000  R      0x1000
199:     LOAD           0x00000000000da000 0x00007fa9cdffd000 0x00007fa9cdffd000
199:                    0x0000000000002000 0x0000000000002000  RW     0x1000
199:     LOAD           0x00000000000dc000 0x00007fa9ce000000 0x00007fa9ce000000
199:                    0x0000000000400000 0x0000000000400000  RW     0x1000
199:     LOAD           0x00000000004dc000 0x00007fa9d2000000 0x00007fa9d2000000
199:                    0x0000000000050000 0x0000000000050000  R      0x1000
199:     LOAD           0x000000000052c000 0x00007fa9d2050000 0x00007fa9d2050000
199:                    0x00000000002f3000 0x00000000002f3000  R E    0x1000
199:     LOAD           0x000000000081f000 0x00007fa9d2343000 0x00007fa9d2343000
199:                    0x00000000000c9000 0x00000000000c9000  R      0x1000
199:     LOAD           0x00000000008e8000 0x00007fa9d240c000 0x00007fa9d240c000
199:                    0x0000000000082000 0x0000000000082000  RW     0x1000
199:     LOAD           0x000000000096a000 0x00007fa9d248e000 0x00007fa9d248e000
199:                    0x000000000003a000 0x000000000003a000  RW     0x1000
199:     LOAD           0x00000000009a4000 0x00007fa9d24fd000 0x00007fa9d24fd000
199:                    0x0000000000001000 0x0000000000001000  R      0x1000
199:     LOAD           0x00000000009a5000 0x00007fa9d24fe000 0x00007fa9d24fe000
199:                    0x000000000002b000 0x000000000002b000  R E    0x1000
199:     LOAD           0x00000000009d0000 0x00007fa9d2529000 0x00007fa9d2529000
199:                    0x000000000000a000 0x000000000000a000  R      0x1000
199:     LOAD           0x00000000009da000 0x00007fa9d2533000 0x00007fa9d2533000
199:                    0x0000000000004000 0x0000000000004000  RW     0x1000
199:     LOAD           0x00000000009de000 0x00007fa9d253d000 0x00007fa9d253d000
199:                    0x0000000000000000 0x0000000000004000  R      0x1000
199:     LOAD           0x00000000009de000 0x00007fa9d2541000 0x00007fa9d2541000
199:                    0x0000000000002000 0x0000000000002000  R      0x1000
199:     LOAD           0x00000000009e0000 0x00007fa9d2543000 0x00007fa9d2543000
199:                    0x0000000000002000 0x0000000000002000  R E    0x1000
199:     LOAD           0x00000000009e2000 0x00007ffced22e000 0x00007ffced22e000
199:                    0x0000000000022000 0x0000000000022000  RW     0x1000
199: 
199:   
199: 
199:   There is no dynamic section in this file.
199: 
199: 
199: 
199:   Displaying notes found at file offset 0x00000af8 with length 0x00000378:
199: 
199:     Owner                Data size 	Description
199:     CORE                 0x00000150	NT_PRSTATUS (prstatus structure)
199:     CORE                 0x00000200	NT_FPREGSET (floating point registers)
199: 
199:   | failed to match expected output |thank you for testing memory dump
199: 
199:   thread init
199: 
199:   nudge delivered 1
199: 
199:   Hello, world!
199: 
199:   .*ELF Header:
199: 
199:     Magic:   7f 45 4c 46 02 01 01 03 00 00 00 00 00 00 00 00.*
199:     Class:                             ELF64
199:     Data:                              2's complement, little endian
199:     Version:                           1 \(current\)
199:     OS/ABI:                            UNIX - GNU
199:     ABI Version:                       0
199:     Type:                              CORE \(Core file\)
199:     Machine:                           .*
199:     Version:                           0x1
199:     Entry point address:               0x[0-9a-f]+
199:     Start of program headers:          64 \(bytes into file\)
199:     Start of section headers:          [0-9]+ \(bytes into file\)
199:     Flags:                             0x0
199:     Size of this header:               64 \(bytes\)
199:     Size of program headers:           56 \(bytes\)
199:     Number of program headers:         [0-9]+
199:     Size of section headers:           64 \(bytes\)
199:     Number of section headers:         [0-9]+
199:     Section header string table index: [0-9]+
199: 
199:   .*
199: 
199:   Section Headers:
199: 
199:   .*\[Nr\] Name Type Address Offset
199: 
199:          Size              EntSize          Flags  Link  Info  Align
199: 
199:   .* \[vvar\] PROGBITS.*
199: 
199:   .* \[vdso\] PROGBITS.*
199: 
199:   .* shstrtab STRTAB.*
199: 
199:   Program Headers:
199: 
199:     Type           Offset             VirtAddr           PhysAddr
199:                    FileSiz            MemSiz              Flags  Align
199:     NOTE.*
199:     LOAD.*
199: 
199:   .*
199: 
199:   Displaying notes found at file offset 0x[0-9a-f]+ with length 0x[0-9a-f]+:
199: 
199:     Owner.*Data size.*Description
199:     CORE.*NT_PRSTATUS.*\(prstatus structure\)
199:     CORE.*NT_FPREGSET.*\(floating point registers\)
199: 
199:   |
199: 
199: 
1/1 Test #199: code_api|client.memory_dump_test ...***Failed    0.23 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.29 sec

The following tests FAILED:
	199 - code_api|client.memory_dump_test (Failed)
Errors while running CTest
Output from these tests are in: /home/binwang/projects/dynamorio/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
```
</details>

The relevant part seems to be
```
199:   <Failed to write the requested memory content into the core dump file.>
199: 
199:   <section: [vvar_vclock], prot: 1, length: 8192, written: -14
```

--- a/core/unix/coredump.c
+++ b/core/unix/coredump.c
@@ -58,7 +58,7 @@
 #define MAX_SECTION_HEADERS 300
 #define MAX_SECTION_NAME_BUFFER_SIZE 8192
 #define SECTION_HEADER_TABLE ".shstrtab"
-#define VVAR_SECTION "[vvar]"
+#define VVAR_SECTION_PREFIX "[vvar"
 #define VSYSCALL_SECTION "[vsyscall]"
 /*
  * The length of the name has to be a multiple of eight (for X64) to ensure the next
@@ -95,6 +95,8 @@ typedef struct _section_header_info_t {
     app_pc vm_end;
     uint prot;
     ELF_ADDR name_offset;
+    bool is_vvar; /* true for [vvar] and [vvar_vclock]: kernel-owned pages that fault on
+                     direct access despite PROT_READ */
 } section_header_info_t;
 
 /*
@@ -461,17 +463,16 @@ os_dump_core_internal(dcontext_t *dcontext, const char *output_directory DR_PARA
     // in iter.comment. Region names are stored in the section name table
     // without duplications. An offset is used in the section header to locate
     // the section name in the section name table.
-    int vvar_section = -1;
     while (memquery_iterator_next(&iter)) {
         // Skip non-readable sections during processing, with the exception of the VVAR
-        // section. The VVAR mapping is included in the core dump file, but its contents
-        // are not saved.
+        // sections. The VVAR mappings are included in the core dump file, but their
+        // contents are not saved.
         if (iter.prot == MEMPROT_NONE || strcmp(iter.comment, VSYSCALL_SECTION) == 0) {
             continue;
         }
-        if (strcmp(iter.comment, VVAR_SECTION) == 0) {
-            vvar_section = section_count;
-        }
+        /* [vvar] and [vvar_vclock] are kernel-owned pages that return EFAULT on read. */
+        const bool is_vvar = strncmp(iter.comment, VVAR_SECTION_PREFIX,
+                                     sizeof(VVAR_SECTION_PREFIX) - 1) == 0;
         ELF_ADDR offset = 0;
         if (iter.comment != NULL && iter.comment[0] != '\0') {
             TABLE_RWLOCK(string_htable, write, lock);
@@ -498,7 +499,8 @@ os_dump_core_internal(dcontext_t *dcontext, const char *output_directory DR_PARA
         section_header_info[section_count].vm_end = iter.vm_end;
         section_header_info[section_count].prot = iter.prot;
         section_header_info[section_count].name_offset = offset;
-        if (section_count != vvar_section) {
+        section_header_info[section_count].is_vvar = is_vvar;
+        if (!is_vvar) {
             section_data_size += iter.vm_end - iter.vm_start;
         }
         ++section_count;
@@ -600,13 +602,13 @@ os_dump_core_internal(dcontext_t *dcontext, const char *output_directory DR_PARA
                 /*virtual_address=*/(ELF_ADDR)section_header_info[section_index].vm_start,
                 /*physical_address=*/
                 (ELF_ADDR)section_header_info[section_index].vm_start,
-                /*file_size=*/section_index == vvar_section ? 0 : size,
+                /*file_size=*/section_header_info[section_index].is_vvar ? 0 : size,
                 /*memory_size=*/size,
                 /*alignment=*/os_page_size())) {
             os_close(elf_file);
             return false;
         }
-        if (section_index != vvar_section) {
+        if (!section_header_info[section_index].is_vvar) {
             core_file_offset += section_header_info[section_index].vm_end -
                 section_header_info[section_index].vm_start;
         }
@@ -643,10 +645,9 @@ os_dump_core_internal(dcontext_t *dcontext, const char *output_directory DR_PARA
     }
     // Write memory content to the core dump file.
     for (int section_index = 0; section_index < section_count - 1; ++section_index) {
-        // The vvar section is not readable. The program header has the file
-        // size set to zero to indicate the content is not recorded in the
-        // memory dump file.
-        if (section_index == vvar_section) {
+        // [vvar] and [vvar_vclock] are not readable. Their program headers have
+        // file size zero, so there is no content to write.
+        if (section_header_info[section_index].is_vvar) {
             continue;
         }
         const size_t length = section_header_info[section_index].vm_end -
@@ -700,7 +701,7 @@ os_dump_core_internal(dcontext_t *dcontext, const char *output_directory DR_PARA
             os_close(elf_file);
             return false;
         }
-        if (section_index != vvar_section) {
+        if (!section_header_info[section_index].is_vvar) {
             core_file_offset += section_header_info[section_index].vm_end -
                 section_header_info[section_index].vm_start;
         }

--- a/core/unix/coredump.c
+++ b/core/unix/coredump.c
@@ -94,8 +94,9 @@ typedef struct _section_header_info_t {
     app_pc vm_end;
     uint prot;
     ELF_ADDR name_offset;
-    bool is_vvar; /* true for [vvar] and [vvar_vclock]: kernel-owned pages that fault on
-                     direct access despite PROT_READ */
+    /* This field is true for [vvar] and [vvar_vclock]. They are kernel-owned pages that
+     * fault on direct access despite PROT_READ. */
+    bool is_vvar;
 } section_header_info_t;
 
 /*

--- a/core/unix/coredump.c
+++ b/core/unix/coredump.c
@@ -1,5 +1,5 @@
 /* **********************************************************
- * Copyright (c) 2024 Google, Inc.  All rights reserved.
+ * Copyright (c) 2024-2026 Google, Inc.  All rights reserved.
  * **********************************************************/
 
 /*

--- a/core/unix/coredump.c
+++ b/core/unix/coredump.c
@@ -58,7 +58,6 @@
 #define MAX_SECTION_HEADERS 300
 #define MAX_SECTION_NAME_BUFFER_SIZE 8192
 #define SECTION_HEADER_TABLE ".shstrtab"
-#define VVAR_SECTION_PREFIX "[vvar"
 #define VSYSCALL_SECTION "[vsyscall]"
 /*
  * The length of the name has to be a multiple of eight (for X64) to ensure the next
@@ -471,8 +470,8 @@ os_dump_core_internal(dcontext_t *dcontext, const char *output_directory DR_PARA
             continue;
         }
         /* [vvar] and [vvar_vclock] are kernel-owned pages that return EFAULT on read. */
-        const bool is_vvar = strncmp(iter.comment, VVAR_SECTION_PREFIX,
-                                     sizeof(VVAR_SECTION_PREFIX) - 1) == 0;
+        const bool is_vvar = strncmp(iter.comment, VVAR_PAGE_MAPS_NAME_PREFIX,
+                                     strlen(VVAR_PAGE_MAPS_NAME_PREFIX)) == 0;
         ELF_ADDR offset = 0;
         if (iter.comment != NULL && iter.comment[0] != '\0') {
             TABLE_RWLOCK(string_htable, write, lock);

